### DISCUSSION TITLE
LVGL image some times replaced with 'No data'

### DISCUSSION THREAD

**Discussion by ukoda** [arendst/Tasmota#20390] [Support request] [answered]
I am having trouble in a Berry LVGL scirpt with a status image sometimes being replaced with a white box containing the text 'No data'.  You can see more detail of the project in the now resolved issue (https://github.com/arendst/Tasmota/issues/20350).  The original C code was created with the aid of a GUI layout tool so I was using an image button and toggling between the two status images by calling set_src function to read the png image from the file system.  Once the image is replaced with 'No data' it will never appear again until the program is restarted.  I have not found a relaiable to force the failure so have not created a bug report yet.

When I saw this problem I figure it was time to better implement the status image.  I replaced the image button with two images and use the lv.OBJ_FLAG_HIDDEN flag to toggle between them.  The intent was to hold the images in RAM instead of reading and decoding them from flash at each change.  It worked ok at first but now has the same problem.

It looks like you can't attach a Berry file here so here it is in full.  Sorry, the indenting formats wrong in markdown, some evil tab characters have found there way into my source code making my code look like a mess here.  I will banish them later and tell my editor to never insert evil tab characters again. The initial load of the images occurs at lines 105 to 128.  The toggle of the images occurs at lines 193 to 199.
```
#-----------------------------------------------------------------------------

  EMA Panel Heater

-----------------------------------------------------------------------------#

import json
import persist

# Set state vars

var Mode            = "Unknown"   # Flag as unklnown so we can ensure it is remote mode at end of init process
var TemperatureRoom = 20
var TemperatureSet  = 24
var Element         = true        # Flag as on so we can ensure it is turned off at end of init process
var Tick            = 1

if persist.has("TemperatureSet")
  TemperatureSet = int(persist.member("TemperatureSet"))
  print("Restored temperature set as", TemperatureSet)
end

# Start LVGL and init environment

lv.start()
scr = lv.scr_act()
scr.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
scr.set_style_bg_color(lv.color(0x000000), lv.PART_MAIN | lv.STATE_DEFAULT)
var lg_font = lv.font_montserrat(28)

# Mode panel

ui_ModePanel = lv.obj(scr)
ui_ModePanel.set_width(130)
ui_ModePanel.set_height(220)
ui_ModePanel.set_x(10)
ui_ModePanel.set_y(0)
ui_ModePanel.set_align(lv.ALIGN_LEFT_MID)
ui_ModePanel.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
ui_ModePanel.set_style_bg_color(lv.color(0x000000), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_ModePanel.set_style_text_color(lv.color(0xFFFFFF), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_ModePanel.set_style_border_color(lv.color(0x787676), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_ModePanel.set_style_border_opa(255, lv.PART_MAIN | lv.STATE_DEFAULT)

ui_RemoteButton = lv.btn(ui_ModePanel)
ui_RemoteButton.set_width(100)
ui_RemoteButton.set_height(50)
ui_RemoteButton.set_x(0)
ui_RemoteButton.set_y(-68)
ui_RemoteButton.set_align(lv.ALIGN_CENTER)
ui_RemoteButton.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
ui_RemoteButton.set_style_outline_color(lv.color_hex(0xFFFEFE), lv.STATE_USER_1)
ui_RemoteButton.set_style_outline_opa(255, lv.STATE_USER_1)
ui_RemoteButton.set_style_outline_width(2, lv.STATE_USER_1)
ui_RemoteButton.set_style_outline_pad(2, lv.STATE_USER_1)

ui_RemoteButtonLabel = lv.label(ui_RemoteButton)
ui_RemoteButtonLabel.set_width(lv.SIZE_CONTENT)
ui_RemoteButtonLabel.set_height(lv.SIZE_CONTENT)
ui_RemoteButtonLabel.set_align(lv.ALIGN_CENTER)
ui_RemoteButtonLabel.set_text("Remote")

ui_OnButton = lv.btn(ui_ModePanel)
ui_OnButton.set_width(100)
ui_OnButton.set_height(50)
ui_OnButton.set_align(lv.ALIGN_CENTER)
ui_OnButton.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
ui_OnButton.set_style_outline_width(2, lv.PART_MAIN | lv.STATE_DEFAULT)
ui_OnButton.set_style_outline_pad(2, lv.PART_MAIN | lv.STATE_DEFAULT)
ui_OnButton.set_style_text_color(lv.color_hex(0xFFFFFF), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_OnButton.set_style_text_opa(255, lv.PART_MAIN | lv.STATE_DEFAULT)
ui_OnButton.set_style_outline_color(lv.color_hex(0xFFFEFE), lv.STATE_USER_1)
ui_OnButton.set_style_outline_opa(255, lv.STATE_USER_1)
ui_OnButton.set_style_outline_width(2, lv.STATE_USER_1)
ui_OnButton.set_style_outline_pad(2, lv.STATE_USER_1)

ui_OnButtonLabel = lv.label(ui_OnButton)
ui_OnButtonLabel.set_width(lv.SIZE_CONTENT)
ui_OnButtonLabel.set_height(lv.SIZE_CONTENT)
ui_OnButtonLabel.set_align(lv.ALIGN_CENTER)
ui_OnButtonLabel.set_text("On")

ui_OffButton = lv.btn(ui_ModePanel)
ui_OffButton.set_width(100)
ui_OffButton.set_height(50)
ui_OffButton.set_x(0)
ui_OffButton.set_y(68)
ui_OffButton.set_align(lv.ALIGN_CENTER)
ui_OffButton.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
ui_OffButton.set_style_text_color(lv.color_hex(0xFFFFFF), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_OffButton.set_style_text_opa(255, lv.PART_MAIN | lv.STATE_DEFAULT)
ui_OffButton.set_style_outline_color(lv.color_hex(0xFFFEFE), lv.STATE_USER_1)
ui_OffButton.set_style_outline_opa(255, lv.STATE_USER_1)
ui_OffButton.set_style_outline_width(2, lv.STATE_USER_1)
ui_OffButton.set_style_outline_pad(2, lv.STATE_USER_1)

ui_OffButtonLabel = lv.label(ui_OffButton)
ui_OffButtonLabel.set_width(lv.SIZE_CONTENT)
ui_OffButtonLabel.set_height(lv.SIZE_CONTENT)
ui_OffButtonLabel.set_align(lv.ALIGN_CENTER)
ui_OffButtonLabel.set_text("Off")

# Status panel

ui_StatePanel = lv.obj(scr)
ui_StatePanel.set_width(160)
ui_StatePanel.set_height(105)
ui_StatePanel.set_x(-10)
ui_StatePanel.set_y(10)
ui_StatePanel.set_align(lv.ALIGN_TOP_RIGHT)
ui_StatePanel.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
ui_StatePanel.set_style_bg_color(lv.color(0x000000), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_StatePanel.set_style_border_color(lv.color_hex(0x787676), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_StatePanel.set_style_border_opa(255, lv.PART_MAIN | lv.STATE_DEFAULT)

ui_Sleep_Img = lv.img(ui_StatePanel)
ui_Sleep_Img.set_src("A:/sleeping113x80.png")
ui_Sleep_Img.set_width(113)
ui_Sleep_Img.set_height(80)
ui_Sleep_Img.set_align(lv.ALIGN_CENTER)

ui_Heat_Img = lv.img(ui_StatePanel)
ui_Heat_Img.set_src("A:/campfire113x80.png")
ui_Heat_Img.set_width(113)
ui_Heat_Img.set_height(80)
ui_Heat_Img.set_align(lv.ALIGN_CENTER)
ui_Heat_Img.add_flag(lv.OBJ_FLAG_HIDDEN)

# Temperature panel

ui_TemperaturePanel = lv.obj(scr)
ui_TemperaturePanel.set_width(160)
ui_TemperaturePanel.set_height(105)
ui_TemperaturePanel.set_x(-10)
ui_TemperaturePanel.set_y(-10)
ui_TemperaturePanel.set_align(lv.ALIGN_BOTTOM_RIGHT)
ui_TemperaturePanel.clear_flag(lv.OBJ_FLAG_SCROLLABLE)
ui_TemperaturePanel.set_style_bg_color(lv.color(0x000000), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_TemperaturePanel.set_style_text_color(lv.color(0xFFFFFF), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_TemperaturePanel.set_style_border_color(lv.color(0x787676), lv.PART_MAIN | lv.STATE_DEFAULT)
ui_TemperaturePanel.set_style_border_opa(255, lv.PART_MAIN | lv.STATE_DEFAULT)

ui_SetLabel = lv.label(ui_TemperaturePanel)
ui_SetLabel.set_width(50)
ui_SetLabel.set_height(lv.SIZE_CONTENT)
ui_SetLabel.set_align(lv.ALIGN_BOTTOM_LEFT)
ui_SetLabel.set_text("Set")
ui_SetLabel.set_style_text_align(lv.TEXT_ALIGN_CENTER, lv.PART_MAIN | lv.STATE_DEFAULT)

ui_RoomLabel = lv.label(ui_TemperaturePanel)
ui_RoomLabel.set_width(50)
ui_RoomLabel.set_height(lv.SIZE_CONTENT)
ui_RoomLabel.set_align(lv.ALIGN_BOTTOM_RIGHT)
ui_RoomLabel.set_text("Room")
ui_RoomLabel.set_style_text_align(lv.TEXT_ALIGN_CENTER, lv.PART_MAIN | lv.STATE_DEFAULT)

ui_TempUpButton = lv.imgbtn(ui_TemperaturePanel)
ui_TempUpButton.set_src(lv.IMGBTN_STATE_RELEASED, nil, "A:/RedArrow.png", nil)
ui_TempUpButton.set_width(20)
ui_TempUpButton.set_height(20)
ui_TempUpButton.set_align(lv.ALIGN_TOP_MID)

ui_TempDownButton = lv.imgbtn(ui_TemperaturePanel)
ui_TempDownButton.set_src(lv.IMGBTN_STATE_RELEASED, nil, "A:/DarkBlueArrow.png", nil)
ui_TempDownButton.set_width(20)
ui_TempDownButton.set_height(20)
ui_TempDownButton.set_align(lv.ALIGN_BOTTOM_MID)

ui_SetValueLabel = lv.label(ui_TemperaturePanel)
ui_SetValueLabel.set_width(50)
ui_SetValueLabel.set_height(40)
ui_SetValueLabel.set_x(0)
ui_SetValueLabel.set_y(5)
ui_SetValueLabel.set_text(str(TemperatureSet))
ui_SetValueLabel.set_style_text_align(lv.TEXT_ALIGN_CENTER, lv.PART_MAIN | lv.STATE_DEFAULT)
ui_SetValueLabel.set_style_text_font(lg_font, lv.PART_MAIN | lv.STATE_DEFAULT)

ui_RoomValueLabel = lv.label(ui_TemperaturePanel)
ui_RoomValueLabel.set_width(50)
ui_RoomValueLabel.set_height(40)
ui_RoomValueLabel.set_x(0)
ui_RoomValueLabel.set_y(5)
ui_RoomValueLabel.set_align(lv.ALIGN_TOP_RIGHT)
ui_RoomValueLabel.set_text("--")
ui_RoomValueLabel.set_style_text_align(lv.TEXT_ALIGN_CENTER, lv.PART_MAIN | lv.STATE_DEFAULT)
ui_RoomValueLabel.set_style_text_font(lg_font, lv.PART_MAIN | lv.STATE_DEFAULT)

print('GUI startup complete')


def updateelement(elementon)
  Element = elementon
  if (Element)
    ui_Sleep_Img.add_flag(lv.OBJ_FLAG_HIDDEN)
    ui_Heat_Img.clear_flag(lv.OBJ_FLAG_HIDDEN)
  else 
    ui_Sleep_Img.clear_flag(lv.OBJ_FLAG_HIDDEN)
    ui_Heat_Img.add_flag(lv.OBJ_FLAG_HIDDEN)
  end
end


def setelement(elementon)
  if (Element != elementon)
    updateelement(elementon)
  	if (Element)
	    print("Turning on heating element")
      tasmota.set_power(0, true)
	  else 
	    print("Turning off heating element")
      tasmota.set_power(0, false)
    end
  end
end


def setelementbytemperature()
	setelement(TemperatureRoom < TemperatureSet)
end


def setmode(NewMode)
  if Mode != NewMode
    print("Mode changed from", Mode, "to", NewMode)
    Mode = NewMode
    persist.setmember("OperationMode", Mode)
    if Mode == "Remote"
      setelement(false)
      ui_RemoteButton.add_state(lv.STATE_USER_1)
      ui_OnButton.clear_state(lv.STATE_USER_1)
      ui_OffButton.clear_state(lv.STATE_USER_1)
    elif Mode == "On"
      setelementbytemperature()
      ui_RemoteButton.clear_state(lv.STATE_USER_1)
      ui_OnButton.add_state(lv.STATE_USER_1)
      ui_OffButton.clear_state(lv.STATE_USER_1)
    elif Mode == "Off"
      setelement(false)
      ui_RemoteButton.clear_state(lv.STATE_USER_1)
      ui_OnButton.clear_state(lv.STATE_USER_1)
      ui_OffButton.add_state(lv.STATE_USER_1)
    end
  end  
end


#- callback function when a button is pressed, react to EVENT_CLICKED event -#

def btn_clicked_cb(obj, event)
  if   obj == ui_RemoteButton
    setmode("Remote")
  elif obj == ui_OnButton
    setmode("On")
  elif obj == ui_OffButton
    setmode("Off")
  elif obj == ui_TempUpButton
    TemperatureSet += 1
    if Mode == "On"
      setelementbytemperature()
    end
    ui_SetValueLabel.set_text(str(TemperatureSet))
    persist.setmember("TemperatureSet", str(TemperatureSet))
  elif obj == ui_TempDownButton
    TemperatureSet -= 1
    if Mode == "On"
      setelementbytemperature()
    end
    ui_SetValueLabel.set_text(str(TemperatureSet))
    persist.setmember("TemperatureSet", str(TemperatureSet))
  end
end


def get_room_temperature()
  return TemperatureRoom
end


def set_room_temperature(new_temperature)
  TemperatureRoom = new_temperature
end


def mqtt_mode_cmd(cmd, idx, payload, payload_json)
  print("MQTT mode change to", payload)
  setmode(payload)

  # report the command as successful
  tasmota.resp_cmnd_done()
end


def mqtt_power_cmd(cmd, idx, payload, payload_json)
  print("MQTT power cmd", cmd)
  print("MQTT power idx", idx)
  print("MQTT power payload", payload)
  print("MQTT power payload_json", payload_json)
end


class one_second_update
  def every_second()
    var sensors_json = json.load(tasmota.read_sensors())
    var NewTemperture = int(sensors_json['AM2301']['Temperature'])
    var power_list = tasmota.get_power()

    if get_room_temperature() != NewTemperture
      set_room_temperature(NewTemperture)
      ui_RoomValueLabel.set_text(str(NewTemperture))
      if Mode == "On"
        setelementbytemperature()
      end
    end
    if Element != power_list[0]
      print("Updating element status")
      updateelement(power_list[0])
    end
  end

  def json_append()
    var msg = ",\"OperationMode\":\"" + Mode + "\""
    tasmota.response_append(msg)
  end
end


update_driver = one_second_update()
tasmota.add_driver(update_driver)
tasmota.add_cmd('Mode', mqtt_mode_cmd)
tasmota.add_cmd('Power1', mqtt_power_cmd)
  

if persist.has("OperationMode")
  setmode(persist.member("OperationMode"))
  print("Restored mode as", Mode)
else
  setmode("Remote")
end

ui_RemoteButton.add_event_cb(btn_clicked_cb, lv.EVENT_CLICKED, 0)
ui_OnButton.add_event_cb(btn_clicked_cb, lv.EVENT_CLICKED, 0)
ui_OffButton.add_event_cb(btn_clicked_cb, lv.EVENT_CLICKED, 0)
ui_TempUpButton.add_event_cb(btn_clicked_cb, lv.EVENT_CLICKED, 0)
ui_TempDownButton.add_event_cb(btn_clicked_cb, lv.EVENT_CLICKED, 0)
```

**Comment by sfromis** [comment, marked as answer]
Based on the other thread, it looks like you're using a board without PSRAM available, and this is not recommended for doing things with LVGL or HASPmota. While it may work for smaller things, what you're doing there is a lot more than "small", especially when having multiple images.

Better watch available RAM closely, and do not get "too low". The STATE messages at TelePeriod should have fields like `"Heap":249` and `"Berry":{"HeapUsed":4,"Objects":45}`. The output of command `Status 5` should have data like:
`"Free":1068,"Heap":240,"StackLowMark":3,"PsrMax":8188,"PsrFree":8165`

 And in Berry, `tasmota.memory()` should have something like
`{'stack_low': 3.55469, 'program_free': 1068, 'program': 1811, 'psram': 8188, 'stack_size': 8, 'flash_real': 16384, 'iram_free': 0, 'flash': 16384, 'frag': 7, 'heap_free': 243, 'psram_free': 8170}`

Obviously, the above is from a board with plenty of RAM available, both types.

**Reply by ukoda** [reply]
Thanks for that, it would explain a lot.  I will dig a bit deeper.  I would call it small but I realise that has a lot to do with context.  It was small when it was a C program, but it did not have all the goodness that is Tasmota behind it.  I have spent a lot of hardware effort on this project such investing in high temperature plastics and 3D printing new panels for the heaters that fit these displays.  So I'm invested in getting it to work.  Some of the options that come to mind:

1. Build a trimmed down Tasmota binary without support for hardware not present.  The problem is I guess this will mainly save flash space, not RAM.
2. Use as reduced palette image, but this would probably impact the image quality.
3. Fit a PSRAM chip, such as a [APS6404](https://www.pjrc.com/store/APS6404L_3SQR.pdf), not sure if this would work, would probably depend on which ESP32 pin is connected to the CS line.

**Reply by sfromis** [reply]
Trimming down Tasmota may reduce flash usage, but when you already need "fat" features like uDisplay, LVGL and Berry, tweaking out much RAM may not be very easy, as drivers are generally not using RAM until active, and your build is already reduced in many of the features. Maybe getting rid of https support could make a difference.

While I understand that you like having good image quality, I'd still suggest a test run with low-color images. And of course watch RAM available.

Sunton does also have similar boards with ESP32-S3 and PSRAM, but I'm not sure if this applies to the smallest 2.8" screen.

**Reply by s-hadinger** [reply]
Displaying an image is always a challenge with low memory, because it needs to be decompressed into a full buffer. 

You could try `#define LV_IMG_CACHE_DEF_SIZE   0` in `tasmota_lv_conf.h` to completely disable the caching of images, it will take more CPU but could save you some precious memory.

**Reply by ukoda** [reply]
Thanks for the suggestions.  So far I tried building a slimmer Tasmota build.  I also set LV_USE_FT_CACHE_MANAGER to 0 along with other #undefs.  The flash size is 11k smaller and heap usage appears to be worse!  I have ordered some PSRAM chips to look at adding.

It is worth noting I need to add a #ifndef ahead of LV_IMG_CACHE_DEF_SIZE  in tasmota_lv_config.h so it didn't clash with new definition in user_config_override.h.

It is late here so I'm calling quits for today.

**Reply by s-hadinger** [reply]
Don't bother with flash size, it's not your bottleneck

**Reply by ukoda** [reply]
Looking at the data collected the free memory is stable while it is running ok and remains at similar levels after the image failure, and later when it all locks up.  My guess is the memory heap pool is getting fragmented and a failure occurs when there is no longer big enough convective blocks to allocate suitable block sizes from the heap.  As such I am assuming there is no clean fixes using software.

A non-ideal option, in my case, might be to hold the images preformatted in flash, like was the case with the original C code.  This appears to be how the Tasmota logo is handled, stored in tasmota/tasmota_xdrv_driver/xdrv_52_3_berry_lvgl.ino.  The question for me is if putting the images in flash going to stop the need for larger size heap allocations?

**Reply by Jason2866** [reply]
> Displaying an image is always a challenge with low memory, because it needs to be decompressed into a full buffer.

You need enough RAM. Storing in flash does not solve. Get a board with PSRAM or add a compatible PSRAM chip to your board. How to connect is described in espressif tech docu. Imho the challenge is the mechanic and solder work to add.

**Comment by ukoda** [comment]
I'm going to close this out here.  The lesson learnt is don't bother with LVGL on Tasmota with devices that lacks PSRAM.

I have parts on order that may make these displays usable.  Until they arrive I have switched over to a larger WT32-SC01 that does have PSRAM and it works much better.  The WT32-SC01 also had a number of issues running LVGL code on Tasmota, but I was able resolve and work around those to get a decent end result.  Next task is to design a new mounting plate to fit them in the heater.

My thanks to @s-hadinger, @sfromis and @Jason2866 for the taking the time to give advice.
